event.go

```diff
@@ -0,0 +1,167 @@
+package mesg
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"sync"
+
+	"github.com/mesg-foundation/core/api/core"
+)
+
+var errCannotDecodeEventData = errors.New("cannot decode event data")
+
+// Event is a MESG event.
+type Event struct {
+	Key  string
+	data string
+}
+
+// Data decodes event data into out.
+func (e *Event) Data(out interface{}) error {
+	return json.Unmarshal([]byte(e.data), out)
+}
+
+// EventEmitter is a MESG event emitter.
+type EventEmitter struct {
+	app *Application
+
+	// event is the actual event to listen for.
+	event string
+
+	//eventServiceID is the service id of where event is emitted.
+	eventServiceID string
+
+	// task is the actual task that will be executed.
+	task string
+
+	// taskServiceID is the service id of target task.
+	taskServiceID string
+
+	// filterFuncs holds funcs that returns boolean values to decide
+	// if the task should be executed or not.
+	filterFuncs []func(*Event) bool
+
+	// mapFunc is a func that returns input data of task.
+	mapFunc func(*Event) Data
+
+	// m protects emitter configuration.
+	m sync.Mutex
+
+	// cancel cancels listening for upcoming events.
+	cancel context.CancelFunc
+}
+
+// EventOption is the configuration func of EventListener.
+type EventOption func(*EventEmitter)
+
+// EventFilterOption returns a new option to filter events by name.
+// Default is all(*).
+func EventFilterOption(event string) EventOption {
+	return func(l *EventEmitter) {
+		l.event = event
+	}
+}
+
+// WhenEvent creates an EventListener for serviceID.
+func (a *Application) WhenEvent(serviceID string, options ...EventOption) *EventEmitter {
+	e := &EventEmitter{
+		app:            a,
+		eventServiceID: serviceID,
+		event:          "*",
+	}
+	for _, option := range options {
+		option(e)
+	}
+	return e
+}
+
+// Filter sets filter funcs that will be executed to decide to execute the
+// task or not.
+// It's possible to add multiple filters by calling Filter multiple times.
+// Other filter funcs and the task execution will no proceed if a filter
+// func returns false.
+func (e *EventEmitter) Filter(fn func(*Event) bool) *EventEmitter {
+	e.m.Lock()
+	defer e.m.Unlock()
+	e.filterFuncs = append(e.filterFuncs, fn)
+	return e
+}
+
+// Map sets the returned data as the input data of task.
+// You can dynamically produce input values for task over event data.
+func (e *EventEmitter) Map(fn func(*Event) Data) Executor {
+	e.m.Lock()
+	defer e.m.Unlock()
+	e.mapFunc = fn
+	return e
+}
+
+// Execute starts for listening events and executes task for serviceID with the
+// output data of event or return value of Map if all Filter funcs returned as true.
+func (e *EventEmitter) Execute(serviceID, task string) (*Listener, error) {
```

Comment:
Creating a named basic type would make sense when we've constants with TaskKey type but in our case names are dynamically passed.

I think, using names for basic types are not needed and complicates things for this case. Also, devs are already going to know that second argument should be the task key because we use the name task in the definition of Execute as `task string`.

The other downside of using named basic types is if devs directly don't set task key in the call time of Execute method and pre define task names as constants in their programs they simply can't do a `callTask := "call"` but have to do a type conversion like this `callTask := mesg.TaskKey("call")` which I find this quite unnecessary.

The reason why I used named [Data](https://github.com/ilgooz/go-application/blob/09aeab2/data.go) type for `interface{}` is, I don't want devs to write `interface{}` as their return types while declaring [Map](https://github.com/ilgooz/go-application/blob/09aeab2/event.go#L90) funcs because it looks too generic, and ugly in the code. And they don't even need to do a type conversion while [returning](https://github.com/ilgooz/go-application/blob/4c17bae/examples/quickstart/quickstart/when.go#L27) from the func because Go handles this in the compile time. 

Go also handles type conversions for func arguments too when funcs argument values directly passed to them but not as variables.